pybedlite/overlap_detector.py

```diff
@@ -178,55 +231,75 @@ def from_ucsc(
             ) from exception
 
 
-class OverlapDetector(Iterable[Interval]):
+GenericGenomicSpan = TypeVar("GenericGenomicSpan", bound=Union[GenomicSpan, StrandedGenomicSpan])
```

Comment:
```suggestion
S = TypeVar("S", bound=Union[GenomicSpan, StrandedGenomicSpan])
```

**issue** Could this be made a single uppercase letter?

I would ideally prefer if type variables/parameters could be named as one (or two, maximally) uppercase letters. I think this is more consistent with convention[^1][^2][^3], and improves the clarity of the code. It helps distinguish type parameters from other classes, and makes it more clear when a class is generic. 

[^1]: > By convention, type parameter names are single, uppercase letters. This stands in sharp contrast to the variable [naming](https://docs.oracle.com/javase%2Ftutorial%2F/java/nutsandbolts/variables.html#naming) conventions that you already know about, and with good reason: Without this convention, it would be difficult to tell the difference between a type variable and an ordinary class or interface name.
https://docs.oracle.com/javase%2Ftutorial%2F/java/generics/types.html

[^2]: > For simple type parameters, a single upper-case letter (from the English alphabet) should be used, starting with `A` (this is different than the Java convention of starting with `T`)
https://docs.scala-lang.org/style/naming-conventions.html#type-parameters-generics

[^3]: > To parameterize the types in a new single function, we need to name the type parameter, just as we do for the value parameters to a function. You can use any identifier as a type parameter name. But we’ll use T because, by convention, type parameter names in Rust are short, often just one letter, and Rust’s type-naming convention is UpperCamelCase. Short for type, T is the default choice of most Rust programmers.
https://doc.rust-lang.org/book/ch10-01-syntax.html

Reply:
A little, yes.

I think the documentation in the latter case could be made clearer by 1) excluding the type variable from the list of classes defined in the module and 2) making it clear that `OverlapDetector` is generic (which as far as I can tell is not currently apparent in the signature or docstring).

e.g.

> `class pybedlite.overlap_detector.OverlapDetector[S](intervals: Iterable[S] | None = None)`
>    Detects and returns overlaps between a set of genomic regions and another genomic region.


fwiw I don't think the average Python user thinks about generics _at all_, and I think it would be preferable to use established conventions for generic typing

Reply:
I guess what I am trying to say is that I find the signature `OverlapDetector(intervals: Iterable[GenericGenomicSpan])` to be less clear than `OverlapDetector[S](intervals: Iterable[S])`, because the former does not make it clear that `OverlapDetector` is actually generic

Reply:
> 1) excluding the type variable from the list of classes defined in the module

Because we don't have this project setup where `__init__.py` defines the namespace the only way I could achieve this in this PR is by making the type variable private with `_X` (or eliminating the docstring so Sphinx doesn't autodoc it, but that is not favorable), but then it would show up in the documented type signature of the OverlapDetector as a private type variable. So I think for now it should be public until the project structure changes.

> 2) making it clear that OverlapDetector is generic (which as far as I can tell is not currently apparent in the signature or docstring).

We don't use the phrase generic, but what the OverlapDetector can contain is documented. I can certainly enhance the docs further to use the phrase "generic" too.

https://github.com/fulcrumgenomics/pybedlite/blob/74f7d031f48036fd2aa2646b17d4f375669e36fe/pybedlite/overlap_detector.py#L251-L252

> I guess what I am trying to say is that I find the signature OverlapDetector(intervals: Iterable[GenericGenomicSpan]) to be less clear than OverlapDetector[S](intervals: Iterable[S]), because the former does not make it clear that OverlapDetector is actually generic

The former is more specifically `OverlapDetector[GenericGenomicSpan](intervals: Iterable[GenericGenomicSpan])` and the word "generic" does appear in the type variable name. At least on that point, I don't agree that is all that unclear.

Reply:
^ I bet Sphinx will someday support Py3.12 syntax which would be the specific form:

- `OverlapDetector[GenericGenomicSpan](intervals: Iterable[GenericGenomicSpan])`

I doubt it is capable of that now, though.

Reply:
Let's use Span here...
```suggestion
Span = TypeVar("GenericGenomicSpan", bound=Union[GenomicSpan, StrandedGenomicSpan])
```

Reply:
That is worse IMO. 

It creates the implication that there exists a hierarchy `Span > GenomicSpan > StrandedGenomicSpan`, which can only increase confusion.

Given the voiced preference for a descriptive name for type variables, I would be fine leaving as `GenericGenomicSpan` or renaming to `GenomicSpanType` (slight preference for the latter, but we will defer to you)

Reply:
Agreed, `SpanType` is clearer: https://github.com/fulcrumgenomics/pybedlite/pull/34/files#r1700567057